cirq/contrib/qcircuit/qcircuit_diagram.py

```diff
@@ -47,15 +48,39 @@ def _render(diagram: circuits.TextDiagramDrawer) -> str:
 
     diagram2 = circuits.TextDiagramDrawer()
     for y in range(h):
+        # have a none wire item on this row
+        row_has_item = False
+        # current horizontal line has something that need to be connected
         for x in range(max(0, w - 1)):
             key = (x, y)
             diagram_text = diagram.entries.get(key)
-            v = '&' + (diagram_text.text if diagram_text else  '') + ' '
+            v = '&' + (diagram_text.text if diagram_text else '') + ' '
+            if diagram_text:
+                row_has_item = True
             diagram2.write(2*x + 1, y, v)
-            post1 = r'\qw' if key in qw else ''
-            post2 = r'\qwx' if key in qwx else ''
-            diagram2.write(2*x + 2, y, post1 + post2)
-        diagram2.write(2*w - 1, y, r'&\qw\\')
+            # if has none wire before, should draw wire
+            post1 = r'\qw' if key in qw and row_has_item else ''
+            post2 = ''
+            if key in qwx:
+                # see if the item in row below can be connected by line
+                lower_item = diagram.entries.get((x, y + 1))
+                # see if the upper row needs to be connected
+                upper_wire = diagram2.entries.get((2 * x + 2, y - 1))
+                upper_wire_t = (upper_wire.text
+                                if upper_wire is not None else '')
+                upper_item = diagram2.entries.get((2 * x + 1, y - 1))
+                upper_item_t = (upper_item.text
+                                if upper_item is not None else '')
+                req_qwx = (r'\qwx' in upper_wire_t or
+                           r'\control' in upper_item_t)
+                if lower_item is not None or req_qwx:
+                    post2 = r'\qwx'
+            diagram2.write(2 * x + 2, y, post1 + post2)
+        # end also needs to be different whether the row has item
+        if row_has_item:
+            diagram2.write(2 * w - 1, y, r'&\qw\\')
+        else:
+            diagram2.write(2 * w - 1, y, r'& \\')  # coverage: ignore
```

Comment:
We should only ignore coverage in very exceptional scenarios where it's impossible/very expensive to test. I don't think this is that case.

Reply:
Thanks for the help! Sorry about the mistake here, I did not understand what the coverage tool was for fully. I suppose I would add the appropriate test into qcircuit_test.py? On a side note, I had some problems with not passing the formatting test. Is there a formatting tool everyone else is using for this project? I ran mypy and pylint with the appropriate configuration file. Neither of them complained about my coding style. Yet the formatting test gave me a hard time.

Reply:
Yes, you should write a new test method in `qcircuit_test.py`. It will contain a test circuit and the expected output similar to this: 

https://github.com/quantumlib/Cirq/blob/d848611149d110c47ab611d7e690e29d1f0adce2/cirq/contrib/qcircuit/qcircuit_test.py#L110-L123

Regarding formatting - you can just run `check/format-incremental --apply` and that will reformat your files properly.

Reply:
btw feel free to jump on https://gitter.im/cirqdev/community if you have questions!

Reply:
Thanks a lot for the help! :+1: I ended up using yapf directly. The ./format-incremental --apply did not really work for me. Perhaps it would be nice for me to mention that by the end of CONTRIBUTING.md? I felt really bad spamming these commits that fail the format testing. Perhaps it would help future newbie's like me who wants to contribute?